plugin/pkg/client/auth/oidc/oidc.go

```diff
@@ -0,0 +1,168 @@
+package oidc
+
+import (
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/coreos/go-oidc/jose"
+	"github.com/coreos/go-oidc/oauth2"
+	"github.com/coreos/go-oidc/oidc"
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/client/restclient"
+)
+
+func init() {
+	if err := restclient.RegisterAuthProviderPlugin("oidc", newOIDCAuthProvider); err != nil {
+		glog.Fatalf("Failed to register oidc auth plugin: %v", err)
+	}
+}
+
+func newOIDCAuthProvider(_ string, cfg map[string]string, persister restclient.AuthProviderConfigPersister) (restclient.AuthProvider, error) {
+	var certAuthData []byte
+	var err error
+	if cfg["idp-certificate-authority-data"] != "" {
+		certAuthData, err = base64.StdEncoding.DecodeString(cfg["idp-certificate-authority-data"])
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	clientConfig := restclient.Config{
+		TLSClientConfig: restclient.TLSClientConfig{
+			CAFile: cfg["idp-certifiate-authority"],
+			CAData: certAuthData,
+		},
+	}
+
+	trans, err := restclient.TransportFor(&clientConfig)
+	if err != nil {
+		return nil, err
+	}
+	hc := &http.Client{Transport: trans}
+
+	issuer, ok := cfg["idp-issuer-url"]
+	if !ok || issuer == "" {
+		return nil, errors.New("Must provide idp-issuer-url")
+	}
+
+	providerCfg, err := oidc.FetchProviderConfig(hc, strings.TrimSuffix(issuer, "/"))
+	if err != nil {
+		return nil, fmt.Errorf("error fetching provider config: %v", err)
+	}
+
+	scopes := strings.Split(cfg["extra-scopes"], ",")
```

Comment:
Sure - note that this is a WIP, so docs and whatnot are still not there yet. 

The way this works (more docs [here](https://developers.google.com/identity/protocols/CrossClientAuth)) is that you pass a scope of `audience:server:client_id:$SOME_OTHER_CLIENT_ID` in the [Authentication Request](http://openid.net/specs/openid-connect-core-1_0.html#AuthRequest). The value of `$SOME_OTHER_CLIENT_ID` is the client for whom you wish to make an ID Token for. In this case, that would be the API Server's client_id. Then, if the authentication requesting client (in our case kubectl) is allowed (as determined by the IdP) to mint tokens for the other client_id, after authentication the ID token it gets back will have the `aud` be the *other* client ID (i.e. API Servers) and the `azp` will be kubectl's client ID.

I've tried to avoid calling this the "azp" flow in favor of "cross-client", because the `azp` field is just some metadata which contains information about who originated the request. The important things to validate remains the signature of the JWT and the `aud`, thus, nothing changes on the server side.

I also avoid mandating, in this PR, the use of the cross-client stuff, as not every IdP implements it. But because it's behavior is triggered simply with the use of a custom `scope`, there doesn't have to be any cross-client specific code - it's just a matter of configuration. I believe this makes it a more generally useful OIDC AuthProvider.

I will, however, as part of this PR document the cross-client flow, and why it is recommended that it, or something which addresses the same security concerns, be used.